src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.TryCloneFile.OSX.cs

```diff
@@ -0,0 +1,87 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Win32.SafeHandles;
+using System.Diagnostics;
+
+namespace System.IO
+{
+    internal static partial class FileSystem
+    {
+        private static bool TryCloneFile(string sourceFullPath, string destFullPath, bool overwrite)
+        {
+            // Try to clone the file immediately, this will only succeed if the
+            // destination doesn't exist, so we don't worry about locking for this one.
+            int flags = Interop.@libc.CLONE_ACL;
+            retry1:
+            if (Interop.@libc.clonefile(sourceFullPath, destFullPath, flags) == 0)
+            {
+                // Success.
+                return true;
+            }
+            Interop.Error error = Interop.Sys.GetLastError();
+
+            // Some filesystems don't support ACLs, so may fail due to trying to copy ACLs.
+            // This will disable them and allow trying again (a maximum of 1 time).
+            if (flags != 0 && error == Interop.Error.EINVAL)
+            {
+                flags = 0;
+                goto retry1;
+            }
+
+            // Try to delete the destination file if we're overwriting.
+            if (error == Interop.Error.EEXIST && overwrite)
+            {
+                // Delete the destination. This should fail on directories. Get a lock to the dest file to ensure we don't copy onto it when
+                // it's locked by something else, and then delete it. It should also fail if destination == source since it's already locked.
```

Comment:
> It should also fail if destination == source

given that would be data loss, I hope so 😄  .. please check for tests.

Reply:
We have `CopyOntoSelf` test to check this :)

> do we also/should we check upstream of here whether src==dest in terms of literal equal strings? that would be an easy safety measure as well.

If we added this, we would also want another test to check for indirection via symlinks. I think it's simpler to leave as-is (obviously given that we make the test pass), since copying onto self is an invalid use of the API (since it will always throw) effectively anyway.

Reply:
I'm happy to add a test regardless if you want though.

Reply:
I don't follow "to ensure we don't copy onto it when it's locked by something else". .. in such a case, we won't copy onto it,  because we can't.
The only reason I see to get the handle here is to prevent deleting the source?

Reply:
> I don't follow "to ensure we don't copy onto it when it's locked by something else". .. in such a case, we won't copy onto it, because we can't.
> The only reason I see to get the handle here is to prevent deleting the source?

It can also be if something else is locking it. We have this for compatibility, this is the main reason we have this code. It also happens to detect if we're copying onto ourselves.